pydeseq2/dds.py

```diff
@@ -213,55 +226,153 @@ def __init__(
             raise ValueError(
                 "Either adata or both counts and metadata arguments must be provided."
             )
-
-        # Convert design_factors to list if a single string was provided.
-        self.design_factors = (
-            [design_factors] if isinstance(design_factors, str) else design_factors
-        )
-        self.continuous_factors = continuous_factors
-
-        if self.obs[self.design_factors].isna().any().any():
-            raise ValueError("NaNs are not allowed in the design factors.")
-        self.obs[self.design_factors] = self.obs[self.design_factors].astype(str)
+        for col in self.obs.columns:
+            if col.strip() != col:
+                raise ValueError("Columns cannot contain leading or trailing spaces")
 
         # Check that design factors don't contain underscores. If so, convert them to
         # hyphens.
-        if np.any(["_" in factor for factor in self.design_factors]):
+        self.design_factors = design_factors
+
+        # Following lines handle the case where the user provides a formula
+        # of the type formulaic can recognize or a column
+        if isinstance(self.design_factors, str):
+            self.continuous_factors = continuous_factors
+            if self.design_factors not in self.obs.columns:
+                # TODO uncomment when we support all of formulaic grammar
+                # try:
+                #     Formula(self.design_factors)
+                # except FormulaInvalidError as err:
+                #     raise (
+                #         f"Design factor {self.design_factors} is not a valid formula"
+                #     ) from err
+                # Remove potential spaces around + signs
+                self.design_factors = re.sub(r"\ \+|\+\ ", "+", self.design_factors)
+                # Check that we have something like ~a + b + a:b:c:d
```

Comment:
Can we also handle general formulae (not restricted to + and :)?

https://matthewwardrop.github.io/formulaic/guides/grammar/